lua/lean/abbreviations.lua

```diff
@@ -33,18 +37,150 @@ local function snippets_nvim_enable(snippets, lean_abbreviations)
+      if text == M.leader then
+        M._clear_abbr_mark()
+        local tmp_extmark = vim.api.nvim_buf_set_extmark(0, abbr_mark_ns, row1, col1,
+          { end_line = row2, end_col = col2 })
+        return vim.schedule(function()
+          row1, col1, row2, col2 = get_extmark_range(abbr_mark_ns, tmp_extmark)
+          if not row1 then return end
+          vim.api.nvim_buf_del_extmark(0, abbr_mark_ns, tmp_extmark)
+          vim.api.nvim_buf_set_text(0, row1, col1, row2, col2, {})
+        end)
+      end
+    end
+  end
+
+  if not M.abbr_mark and char == M.leader then
+    local row, col = unpack(vim.api.nvim_win_get_cursor(0))
+    row = row - 1
+    M.abbr_mark = vim.api.nvim_buf_set_extmark(0, abbr_mark_ns, row, col, {
+      hl_group = 'leanAbbreviationMark',
+      end_line = row,
+      end_col = col,
+      right_gravity = false,
+      end_right_gravity = true,
+    })
+    -- override only for the duration of the abbreviation (clashes with autocompletion plugins)
+    vim.api.nvim_buf_set_keymap(0, 'i', '<CR>', 'v:lua._lean_abbreviations_enter_expr()', {expr = true, noremap = true})
+    vim.api.nvim_buf_set_keymap(0, 'i', '<Tab>', 'v:lua._lean_abbreviations_tab_expr()', {expr = true, noremap = true})
+    return
+  end
+end
+
+function _G._lean_abbreviations_enter_expr()
+  M.convert(true)
+  return '\n'
+end
+
+function _G._lean_abbreviations_tab_expr()
+  M.convert(true)
+  return ' '
+end
+
+function M.convert_abbrev(abbrev)
+  if abbrev:find(M.leader) ~= 1 then return abbrev end
+  abbrev = abbrev:sub(#M.leader + 1)
+  if abbrev:find(M.leader) == 1 then
+    return M.leader .. M.convert_abbrev(abbrev:sub(#M.leader + 1))
+  end
+  local matchlen, fromlen, repl = 0, 99999, ""
+  for from, to in pairs(M.abbreviations) do
+    local curmatchlen = 0
+    for i = 1, math.min(#abbrev, #from) do
+      if abbrev:byte(i) == from:byte(i) then
+        curmatchlen = i
+      else
+        break
+      end
+    end
+    if curmatchlen > matchlen or (curmatchlen == matchlen and #from < fromlen) then
+      matchlen, fromlen, repl = curmatchlen, #from, to
+    end
+  end
+  if matchlen == 0 then return M.leader .. abbrev end
+  return repl .. M.convert_abbrev(abbrev:sub(matchlen + 1))
+end
+
+function M.convert(needs_schedule)
+  if not M.abbr_mark then return end
+  local row1, col1, row2, col2 = get_extmark_range(abbr_mark_ns, M.abbr_mark)
+  M._clear_abbr_mark()
+  if not row1 then return end
+
+  local tmp_extmark = vim.api.nvim_buf_set_extmark(0, abbr_mark_ns, row1, col1,
+    { end_line = row2, end_col = col2 })
+  local conv = function()
+    row1, col1, row2, col2 = get_extmark_range(abbr_mark_ns, tmp_extmark)
+    if not row1 or row1 ~= row2 then return end
+    vim.api.nvim_buf_del_extmark(0, abbr_mark_ns, tmp_extmark)
+    local text = vim.api.nvim_buf_get_lines(0, row1, row1+1, true)[1]:sub(col1 + 1, col2)
+    vim.api.nvim_buf_set_text(0, row1, col1, row2, col2, {M.convert_abbrev(text)})
+  end
```

Comment:
We could add support for $CURSOR here right by setting the cursor to the right spot after checking if it's in the expanded value? Is that something you care about? Otherwise it's missing for compe too (#25) so we can always do it later.